File: exp3.py
class Heap:
    def __init__(self, is_min_heap=True):
        self.heap = []
        self.is_min_heap = is_min_heap 
    def _compare(self, parent, child):
        """Compare function to handle Min-Heap or Max-Heap logic."""
        if self.is_min_heap:
            return parent > child 
        else:
            return parent < child 

    def _heapify_up(self, index):
        """Restore heap property moving upwards."""
        parent_index = (index - 1) // 2
        if index > 0 and self._compare(self.heap[parent_index], self.heap[index]):
            # Swap if parent violates the heap property
            self.heap[parent_index], self.heap[index] = self.heap[index], self.heap[parent_index]
            self._heapify_up(parent_index)

    def _heapify_down(self, index):
        """Restore heap property moving downwards."""
        left_child = 2 * index + 1
        right_child = 2 * index + 2
        smallest_or_largest = index

        # Compare left child
        if left_child < len(self.heap) and self._compare(self.heap[smallest_or_largest], self.heap[left_child]):
            smallest_or_largest = left_child

        # Compare right child
        if right_child < len(self.heap) and self._compare(self.heap[smallest_or_largest], self.heap[right_child]):
            smallest_or_largest = right_child

        # If heap property is violated, swap and continue heapifying down
        if smallest_or_largest != index:
            self.heap[index], self.heap[smallest_or_largest] = self.heap[smallest_or_largest], self.heap[index]
            self._heapify_down(smallest_or_largest)

    def insert(self, value):
        """Insert a value into the heap."""
        self.heap.append(value)
        self._heapify_up(len(self.heap) - 1)

    def delete(self, value):
        """Delete an element from the heap."""
        try:
            index = self.heap.index(value)
            # Replace the element to delete with the last element
            self.heap[index] = self.heap[-1]
            self.heap.pop()
            # Restore heap property
            if index < len(self.heap):
                self._heapify_down(index)
                self._heapify_up(index)
        except ValueError:
            print(f"Value {value} not found in the heap.")

    def display(self):
        """Display the heap."""
        print("Heap contents:", self.heap)

# Example Usage
if __name__ == "__main__":
    print("Min-Heap Operations:")
    min_heap = Heap(is_min_heap=True)
    min_heap.insert(10)
    min_heap.insert(20)
    min_heap.insert(5)
    min_heap.insert(15)
    min_heap.display()

    min_heap.delete(10)
    print("After deleting 10:")
    min_heap.display()

    print("\nMax-Heap Operations:")
    max_heap = Heap(is_min_heap=False)
    max_heap.insert(10)
    max_heap.insert(20)
    max_heap.insert(5)
    max_heap.insert(15)
    max_heap.display()

    max_heap.delete(20)
    print("After deleting 20:")
    max_heap.display()

File: exp3.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_SIZE 100

// Min Heap Functions
void minHeapify(int arr[], int n, int i) {
    int smallest = i;
    int left = 2 * i + 1;
    int right = 2 * i + 2;

    if (left < n && arr[left] < arr[smallest])
        smallest = left;
    if (right < n && arr[right] < arr[smallest])
        smallest = right;

    if (smallest != i) {
        int temp = arr[i];
        arr[i] = arr[smallest];
        arr[smallest] = temp;

        minHeapify(arr, n, smallest);
    }
}

void buildMinHeap(int arr[], int n) {
	int i;
    for (i = n / 2 - 1; i >= 0; i--) {
        minHeapify(arr, n, i);
    }
}

void deleteMin(int arr[], int* n) {
    if (*n <= 0)
        return;

    // Move the last element to the root
    arr[0] = arr[*n - 1];
    (*n)--;

    // Heapify the root element
    minHeapify(arr, *n, 0);
}

// Max Heap Functions
void maxHeapify(int arr[], int n, int i) {
    int largest = i;
    int left = 2 * i + 1;
    int right = 2 * i + 2;

    if (left < n && arr[left] > arr[largest])
        largest = left;
    if (right < n && arr[right] > arr[largest])
        largest = right;

    if (largest != i) {
        int temp = arr[i];
        arr[i] = arr[largest];
        arr[largest] = temp;

        maxHeapify(arr, n, largest);
    }
}

void buildMaxHeap(int arr[], int n) {
	int i;
    for (i = n / 2 - 1; i >= 0; i--) {
        maxHeapify(arr, n, i);
    }
}

void deleteMax(int arr[], int* n) {
    if (*n <= 0)
        return;

    // Move the last element to the root
    arr[0] = arr[*n - 1];
    (*n)--;

    // Heapify the root element
    maxHeapify(arr, *n, 0);
}

// Function to display the heap array
void displayHeap(int arr[], int n) {
	int i;
    for (i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

int main() {
    int minHeap[] = {5, 3, 8, 1, 2, 7, 6};
    int maxHeap[] = {5, 3, 8, 1, 2, 7, 6};
    int nMin = sizeof(minHeap) / sizeof(minHeap[0]);
    int nMax = sizeof(maxHeap) / sizeof(maxHeap[0]);

    // Build Min-Heap
    buildMinHeap(minHeap, nMin);
    printf("Min-Heap before deletion:\n");
    displayHeap(minHeap, nMin);

    // Delete root of Min-Heap (delete min)
    deleteMin(minHeap, &nMin);
    printf("Min-Heap after deleting root:\n");
    displayHeap(minHeap, nMin);

    // Build Max-Heap
    buildMaxHeap(maxHeap, nMax);
    printf("\nMax-Heap before deletion:\n");
    displayHeap(maxHeap, nMax);

    // Delete root of Max-Heap (delete max)
    deleteMax(maxHeap, &nMax);
    printf("Max-Heap after deleting root:\n");
    displayHeap(maxHeap, nMax);

    return 0;
}


